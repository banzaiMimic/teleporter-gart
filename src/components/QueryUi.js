import React from 'react'

import { Teleporter } from '../utils/Teleporter'
import CityGraph from './CityGraph'
import ErrorMessage from './ErrorMessage'
import * as Constants from '../constants'

class QueryUi extends React.Component {

  state = {
    teleporter: new Teleporter(),
    teleporterInput: '',
    output: [],
    graphModel: '',
    showInstructions: false,
    errorLine: ''
  }

  onInputChange(event) {
    this.setState({ teleporterInput: event.target.value })
  }

  submit(event) {
    event.preventDefault()
    this.setState( { 
      output: this.state.teleporter.parseInput(this.state.teleporterInput),
      graphModel: this.state.teleporter.parseGraphData(),
      showInstructions: false,
      errorLine: ''
    }, () => {
      const err = this.state.teleporter.error.hasError
      if(err) {
        this.setState({ 
          showInstructions: true,
          errorLine: this.state.teleporter.error.line
        })
        this.clear()
      }
    } )
  }

  clear() {
    this.state.teleporter.clearData()
    this.setState({
      output: [],
      teleporterInput: '',
      graphModel: null
    })
  }

  loadDefaultInput() {
    this.clear()
    this.setState({
      showInstructions: false,
      teleporterInput: Constants.DEFAULT_INPUT
    })
  }

  render() {
    const output = this.state.output
    const data = this.state.graphModel
    return(
      <>
        <CityGraph data={data}/>
        <h3>Input:</h3>
        {this.state.showInstructions && <ErrorMessage errorLine={this.state.errorLine} />}
        <form>
          <textarea cols={50} rows={16}
            value={this.state.teleporterInput}
            onChange={(e) => this.onInputChange(e)}
          />
          {this.state.output.length === 0 && <input onClick={(e) => this.submit(e)} className="btn-submit" type="submit" value="Submit" />}
        </form>
        <input onClick={() => this.clear()} className="btn-submit" type="submit" value="Clear" />
        <input onClick={() => this.loadDefaultInput()} className="btn-submit" type="submit" value="Load Default Input" />
        
        <h3>Output:</h3>
        <ul>
          {output.map((line, idx) => (
            <li key={idx}>{line}</li>
          ))}
        </ul>
      </>
    )
  }
}

export default QueryUi