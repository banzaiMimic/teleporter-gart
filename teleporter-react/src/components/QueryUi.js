import React from 'react'

import { Teleporter } from './Teleporter'
import CityGraph from './CityGraph'

class QueryUi extends React.Component {

  state = {
    teleporter: new Teleporter(),
    teleporterInput: '',
    output: [],
    graphModel: ''
  }

  onInputChange(event) {
    this.setState({ teleporterInput: event.target.value })
  }

  submit() {
    this.setState( { 
      output: this.state.teleporter.parseInput(this.state.teleporterInput),
      graphModel: this.state.teleporter.parseGraphData()
    } )
  }

  clear() {
    this.state.teleporter.clearData()
    this.setState({
      output: [],
      teleporterInput: '',
      graphModel: null
    })
  }

  loadDefaultInput() {
    this.clear()
    this.setState({
      teleporterInput: `Fortuna - Hemingway
Fortuna - Atlantis
Hemingway - Chesterfield
Chesterfield - Springton
Los Amigos - Paristown
Paristown - Oaktown
Los Amigos - Oaktown
Summerton - Springton
Summerton - Hemingway
cities from Summerton in 1 jumps
cities from Summerton in 2 jumps
can I teleport from Springton to Atlantis
can I teleport from Oaktown to Atlantis
loop possible from Oaktown
loop possible from Fortuna`
    })
  }

  render() {
    const output = this.state.output
    const data = this.state.graphModel
    return(
      <>
        <CityGraph data={data}/>
        <h3>Input:</h3>
        <form>
          <textarea cols={50} rows={16}
            value={this.state.teleporterInput}
            onChange={(e) => this.onInputChange(e)}
          />
          {this.state.output.length === 0 && <input onClick={() => this.submit()} className="btn-submit" type="submit" value="Submit" />}
        </form>
        <input onClick={() => this.clear()} className="btn-submit" type="submit" value="Clear" />
        <input onClick={() => this.loadDefaultInput()} className="btn-submit" type="submit" value="Load Default Input" />
        
        <h3>Output:</h3>
        <ul>
          {output.map((line, idx) => (
            <li key={idx}>{line}</li>
          ))}
        </ul>
      </>
    )
  }
}

export default QueryUi