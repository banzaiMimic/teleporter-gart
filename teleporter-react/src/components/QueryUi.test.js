import React from 'react'
import renderer from 'react-test-renderer'

import QueryUi from './QueryUi'

test('Load Default Input, Submit, Clear functionality snapshot tests', () => {
  const component = renderer.create(<QueryUi />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // trigger loadDefaultInput
  const e = { preventDefault: jest.fn() }
  const btnLoadDefaultInput = component.root.findAllByType('input')[2].props
  const btnSubmit = component.root.findAllByType('input')[0].props
  const btnClear = component.root.findAllByType('input')[1].props

  btnLoadDefaultInput.onClick(e)
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // trigger submit
  btnSubmit.onClick(e)
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // trigger clear
  btnClear.onClick(e)
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

})