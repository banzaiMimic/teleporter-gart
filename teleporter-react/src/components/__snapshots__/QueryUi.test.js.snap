// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load Default Input, Submit, Clear functionality snapshot tests 1`] = `
Array [
  <h3>
    Input:
  </h3>,
  <form>
    <textarea
      cols={50}
      onChange={[Function]}
      rows={16}
      value=""
    />
    <input
      className="btn-submit"
      onClick={[Function]}
      type="submit"
      value="Submit"
    />
  </form>,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Clear"
  />,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Load Default Input"
  />,
  <h3>
    Output:
  </h3>,
  <ul />,
]
`;

exports[`Load Default Input, Submit, Clear functionality snapshot tests 2`] = `
Array [
  <h3>
    Input:
  </h3>,
  <form>
    <textarea
      cols={50}
      onChange={[Function]}
      rows={16}
      value="Fortuna - Hemingway
Fortuna - Atlantis
Hemingway - Chesterfield
Chesterfield - Springton
Los Amigos - Paristown
Paristown - Oaktown
Los Amigos - Oaktown
Summerton - Springton
Summerton - Hemingway
cities from Summerton in 1 jumps
cities from Summerton in 2 jumps
can I teleport from Springton to Atlantis
can I teleport from Oaktown to Atlantis
loop possible from Oaktown
loop possible from Fortuna"
    />
    <input
      className="btn-submit"
      onClick={[Function]}
      type="submit"
      value="Submit"
    />
  </form>,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Clear"
  />,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Load Default Input"
  />,
  <h3>
    Output:
  </h3>,
  <ul />,
]
`;

exports[`Load Default Input, Submit, Clear functionality snapshot tests 3`] = `
Array [
  <div
    className="mermaid"
  >
    graph LR;
1[Fortuna]--&gt;2[Hemingway];
1[Fortuna]--&gt;3[Atlantis];
2[Hemingway]--&gt;1[Fortuna];
2[Hemingway]--&gt;4[Chesterfield];
2[Hemingway]--&gt;9[Summerton];
3[Atlantis]--&gt;1[Fortuna];
4[Chesterfield]--&gt;2[Hemingway];
4[Chesterfield]--&gt;5[Springton];
5[Springton]--&gt;4[Chesterfield];
5[Springton]--&gt;9[Summerton];
6[Los Amigos]--&gt;7[Paristown];
6[Los Amigos]--&gt;8[Oaktown];
7[Paristown]--&gt;6[Los Amigos];
7[Paristown]--&gt;8[Oaktown];
8[Oaktown]--&gt;7[Paristown];
8[Oaktown]--&gt;6[Los Amigos];
9[Summerton]--&gt;5[Springton];
9[Summerton]--&gt;2[Hemingway];

  </div>,
  <h3>
    Input:
  </h3>,
  <form>
    <textarea
      cols={50}
      onChange={[Function]}
      rows={16}
      value="Fortuna - Hemingway
Fortuna - Atlantis
Hemingway - Chesterfield
Chesterfield - Springton
Los Amigos - Paristown
Paristown - Oaktown
Los Amigos - Oaktown
Summerton - Springton
Summerton - Hemingway
cities from Summerton in 1 jumps
cities from Summerton in 2 jumps
can I teleport from Springton to Atlantis
can I teleport from Oaktown to Atlantis
loop possible from Oaktown
loop possible from Fortuna"
    />
  </form>,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Clear"
  />,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Load Default Input"
  />,
  <h3>
    Output:
  </h3>,
  <ul>
    <li>
      cities from Summerton in 1 jumps: Springton,Hemingway
    </li>
    <li>
      cities from Summerton in 2 jumps: Springton,Hemingway,Chesterfield,Fortuna
    </li>
    <li>
      can I teleport from Springton to Atlantis: yes
    </li>
    <li>
      can I teleport from Oaktown to Atlantis: no
    </li>
    <li>
      loop possible from Oaktown: yes
    </li>
    <li>
      loop possible from Fortuna: no
    </li>
  </ul>,
]
`;

exports[`Load Default Input, Submit, Clear functionality snapshot tests 4`] = `
Array [
  <h3>
    Input:
  </h3>,
  <form>
    <textarea
      cols={50}
      onChange={[Function]}
      rows={16}
      value=""
    />
    <input
      className="btn-submit"
      onClick={[Function]}
      type="submit"
      value="Submit"
    />
  </form>,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Clear"
  />,
  <input
    className="btn-submit"
    onClick={[Function]}
    type="submit"
    value="Load Default Input"
  />,
  <h3>
    Output:
  </h3>,
  <ul />,
]
`;
